[Security Solution][Elastic AI Assistant] Adds Knowledge Base

spong opened the pull request:
## Summary

Advanced Settings to enable the Knowledge Base is currently behind the same code toggle introduced in https://github.com/elastic/kibana/pull/164908, please modify `assistantLangChain` to be `true` to enable the Advanced Settings and Knowledge Base options:

https://github.com/elastic/kibana/blob/1dee16e06169a756a3ed8856435ed572fddd20cf/x-pack/plugins/security_solution/public/assistant/provider.tsx#L55

When the above modification is present, a new `Advanced Settings` UI is available within the Assistant that enables the ability to turn on and configure the Knowledge Base, and also add the requisite `ES|QL` Knowledge Base resources.

<p align="center">
  <img width="500" src="https://github.com/elastic/kibana/assets/2946766/3abab422-9a85-45f6-8b0b-854ed7383d1c" />
</p> 

Once enabled, the Assistant will query the Knowledge Base using a `ConversationalRetrievalQAChain`, by means of the ELSER specific `ElasticsearchStore` LangChain `VectorStore` abstraction: 


<p align="center">
  <img width="500" src="https://github.com/elastic/kibana/assets/2946766/a529faab-948c-40f1-bc11-7a39b1766c7f" />
</p> 



### Checklist

Delete any items that are not applicable to this PR.

- [X] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
  * TBD
- [X] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [X] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))

andrew-goldstein commented on x-pack/packages/kbn-elastic-assistant/impl/assistant/settings/advanced_settings/advanced_settings.tsx:
```diff
@@ -5,19 +5,127 @@
+import { useSetupKnowledgeBase } from '../../../knowledge_base/use_setup_knowledge_base/use_setup_knowledge_base';
+import { useDeleteKnowledgeBase } from '../../../knowledge_base/use_delete_knowledge_base/use_delete_knowledge_base';
 
+const ESQL_RESOURCE = 'esql';
```
consider extracting this common type to the package

andrew-goldstein commented on x-pack/packages/kbn-elastic-assistant/impl/assistant/api.tsx:
```diff
@@ -7,12 +7,28 @@
 import { API_ERROR } from './translations';
 import { MODEL_GPT_3_5_TURBO } from '../connectorland/models/model_selector/model_selector';
 
+// TODO: Common imports from plugin
```
TODO: remove this TODO

andrew-goldstein commented on x-pack/packages/kbn-elastic-assistant/impl/assistant/api.tsx:
```diff
@@ -86,3 +102,106 @@ export const fetchConnectorExecuteAction = async ({
+  http,
+  resource,
+  signal,
+}: PostKnowledgeBaseParams): Promise<PostKnowledgeBaseResponse | IHttpFetchError | unknown> => {
```
is `| unknown required`?

andrew-goldstein commented on x-pack/packages/kbn-elastic-assistant/impl/knowledge_base/use_delete_knowledge_base/use_delete_knowledge_base.tsx:
```diff
@@ -0,0 +1,58 @@
+ */
+
+import { useMutation } from '@tanstack/react-query';
+import { IToasts } from '@kbn/core-notifications-browser';
```
nit: import type?

andrew-goldstein commented on x-pack/plugins/elastic_assistant/common/types.ts:
```diff
@@ -0,0 +1,10 @@
+ * 2.0.
+ */
+
+export type { DeleteKnowledgeBaseResponse } from '../server/routes/knowledge_base/delete_knowledge_base';
```
consider removing this after the move to the package

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/content_loaders/esql_loader.ts:
```diff
@@ -0,0 +1,58 @@
+      `Loaded ${response?.length ?? 0} ESQL docs and language docs into the Knowledge Base`
+    );
+
+    return response == null || response.length > 0;
```
conisder adding a comment explaining why response may be `null` or `undefined`, and why we would return true in that case

spong replied:
In fixing this issue (https://github.com/elastic/kibana/pull/166570#issuecomment-1724532953), I went ahead and updated the `addDocuments` return signature to only return a `string[]`, so no need to check or determine if the response is `null` or `undefined`. The number of documents successfully added is always returned.

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/constants.ts:
```diff
@@ -0,0 +1,13 @@
+ * 2.0.
+ */
+
+export const ELSER_MODEL_ID = '.elser_model_1';
```
conisder adding a comment explaining why this should NOT be changed from `.elser_model_1`

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/constants.ts:
```diff
@@ -0,0 +1,13 @@
+export const ELSER_MODEL_ID = '.elser_model_1';
+export const KNOWLEDGE_BASE_INDEX_PATTERN = '.kibana-elastic-ai-assistant-kb';
+export const KNOWLEDGE_BASE_INGEST_PIPELINE = '.kibana-elastic-ai-assistant-kb-ingest-pipeline';
+export const ESQL_DOCS_LOADED_QUERY =
```
consdier adding a comment explaining this content, which is used in a query

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/elasticsearch_store/elasticsearch_store.ts:
```diff
@@ -0,0 +1,316 @@
+    documents: Document[],
+    options?: {}
+  ): Promise<string[] | void> => {
+    // Note: implement if/when needed
```
consider adding a logger info in case langchain starts trying to execute this

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/elasticsearch_store/elasticsearch_store.ts:
```diff
@@ -0,0 +1,316 @@
+    k: number,
+    filter?: this['FilterType']
+  ): Promise<Array<[Document, number]>> => {
+    // Note: Implement if needed
```
consider adding a logger info in case langchain starts trying to execute this

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/elasticsearch_store/elasticsearch_store.ts:
```diff
@@ -0,0 +1,316 @@
+          {
+            text_expansion: {
+              'vector.tokens': {
+                model_id: '.elser_model_1',
```
consider importing this constant (or making it a variable)

spong replied:
Hoisted to class state and included as optional param in constructor ala `index`. Initialized to `this.model = model ?? 'elser_model_1';`

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/elasticsearch_store/elasticsearch_store.ts:
```diff
@@ -0,0 +1,316 @@
+      );
+    } catch (e) {
+      this.logger.error(e);
+      return [];
```
no changes required, but noting that the chain will continue without errors even if a "hard" failure occurs in this step

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/embeddings/elasticsearch_embeddings.ts:
```diff
@@ -0,0 +1,35 @@
+   */
+
+  embedDocuments(documents: string[]): Promise<number[][]> {
+    return Promise.resolve([]);
```
consider `INFO` logs in case LangChain starts calling this and in the `embedQuery` function below

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts:
```diff
@@ -5,47 +5,70 @@
   const pastMessages = langChainMessages.slice(0, -1); // all but the last message
   const latestMessage = langChainMessages.slice(-1); // the last message
-
+  const systemPrompt = langChainMessages[0].content; // the first message
```
consider revisisting this assumption in a future PR per our offline chat

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts:
```diff
@@ -5,47 +5,70 @@
+  // Chain w/ chat history memory and knowledge base retriever
+  const chain = ConversationalRetrievalQAChain.fromLLM(llm, esStore.asRetriever(), {
+    memory,
+    qaChainOptions: { prompt, type: 'stuff' },
```
consider adding a comment re this `type` param

andrew-goldstein replied:
consider removing `prompt`, per desk testing

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts:
```diff
@@ -5,47 +5,70 @@
+  // const chain = RetrievalQAChain.fromLLM(llm, esStore.asRetriever(), {
+  //   prompt,
+  // });
+  // await chain.call({ query: latestMessage[0].content });
```
thanks for keeping this here one more round, as we're still experementing with different chains

spong replied:
`prompt` removed from example as well 👍

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/delete_knowledge_base.ts:
```diff
@@ -0,0 +1,82 @@
+        const esStore = new ElasticsearchStore(esClient, KNOWLEDGE_BASE_INDEX_PATTERN, logger);
+
+        if (kbResource === ESQL_RESOURCE) {
+          // For now, tearing down the Knowledge Base is fine, but will want to support removing specific assets based
```
In the future, consider how the "where clause" to delete the specific assists would be constructed. (Do we need additional metadata in the documents to enable deletion of specific content in the future?)

spong replied:
Commented added to `ESQL_DOCS_LOADED_QUERY` constant with this consideration. Thank you 👍

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/get_knowledge_base_status.ts:
```diff
@@ -0,0 +1,88 @@
+        const pipelineExists = await esStore.pipelineExists();
+        const modelExists = await esStore.isModelInstalled(ELSER_MODEL_ID);
+        const esqlExists =
+          indexExists && (await esStore.similaritySearch(ESQL_DOCS_LOADED_QUERY)).length > 0;
```
consider avoiding this work if some of the conditions above are false

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/get_knowledge_base_status.ts:
```diff
@@ -0,0 +1,88 @@
+
+        const body: GetKnowledgeBaseStatusResponse = {
+          elser_exists: modelExists,
+          esql_exists: kbResource === ESQL_RESOURCE ? esqlExists : undefined,
```
per our offfline chat, hapi will exclude `esql_exists` from the response when this value is `undefined`

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/post_knowledge_base.ts:
```diff
@@ -0,0 +1,88 @@
+          const esqlExists = (await esStore.similaritySearch(ESQL_DOCS_LOADED_QUERY)).length > 0;
+          if (!esqlExists) {
+            const loadedKnowledgeBase = await loadESQL(esStore, logger);
+            return response.ok({ body: { success: loadedKnowledgeBase } });
```
consider returning a `201` here

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/post_knowledge_base.ts:
```diff
@@ -0,0 +1,88 @@
+            const loadedKnowledgeBase = await loadESQL(esStore, logger);
+            return response.ok({ body: { success: loadedKnowledgeBase } });
+          } else {
+            return response.ok({ body: { success: false } });
```
consider returning a `200` with `success: true`, because:
- We are OK if the pipline already exists
- We are OK if the index already exists
- We are OK if the content already exists

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/routes/knowledge_base/post_knowledge_base.ts:
```diff
@@ -0,0 +1,88 @@
+          }
+        }
+
+        return response.ok({ body: { success: indexExists && pipelineExists } });
```
consider returning a `409` or other code if either condition in `indexExists && pipelineExists` is false when we reach this point

andrew-goldstein commented on x-pack/plugins/elastic_assistant/README.md:
```diff
@@ -7,3 +7,9 @@ This plugin does NOT contain UI components. See `x-pack/packages/kbn-elastic-ass
+
+### Testing
+
+To run the tests for this plugin, run `node scripts/jest --watch x-pack/plugins/elastic_assistant/jest.config.js --coverage'
```
consider removing the trailing `'`

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

andrew-goldstein commented:
While desk testing, I observed the following (server) log output, where the total appears to be `130` documents:

```
[2023-09-18T16:06:47.646-06:00][INFO ][plugins.elasticAssistant] Loading 126 ESQL docs and 4 language docs into the Knowledge Base
[2023-09-18T16:07:19.417-06:00][INFO ][plugins.elasticAssistant] Loaded 0 ESQL docs and language docs into the Knowledge Base
```

however per the screenshot from Stack Management > Index Management below, only `129` documents appear in the index:
![stack_mgmt_index_mgmt](https://github.com/elastic/kibana/assets/4459398/7a723555-b8d7-4da7-b37e-bd0869ec5802)

andrew-goldstein commented on x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts:
```diff
@@ -5,47 +5,70 @@
 import { BaseMessage } from 'langchain/schema';
 
-import { ActionsClientLlm } from '../llm/actions_client_llm';
+import { ChatPromptTemplate, SystemMessagePromptTemplate } from 'langchain/prompts';
```
this import may be removed when `prompt` is no longer passed as a param

andrew-goldstein reviewed (approved):
Thanks @spong for the e2e integration of ELSER and LangChain to answer ES|QL knowledge base questions! 🙏 
✅ Desk tested locally
LGTM 🚀

spong commented:
> While desk testing, I observed the following (server) log output, where the total appears to be `130` documents:
> 
> ```
> [2023-09-18T16:06:47.646-06:00][INFO ][plugins.elasticAssistant] Loading 126 ESQL docs and 4 language docs into the Knowledge Base
> [2023-09-18T16:07:19.417-06:00][INFO ][plugins.elasticAssistant] Loaded 0 ESQL docs and language docs into the Knowledge Base
> ```
> 
> however per the screenshot from Stack Management > Index Management below, only `129` documents appear in the index: ![stack_mgmt_index_mgmt](https://user-images.githubusercontent.com/4459398/268785924-7a723555-b8d7-4da7-b37e-bd0869ec5802.png)


Nice catch! So this turned out to be an empty esql doc file, [esql/docs/reference/esql/esql-get-started.asciidoc](https://github.com/elastic/elasticsearch/blob/feature/esql/docs/reference/esql/esql-get-started.asciidoc). I went ahead and removed this file.

I also updated the error/info logger to make it easier to determine which files weren't indexed, and what the errors were, so this should be more helpful next time 🙂

<img width="2157" alt="image" src="https://github.com/elastic/kibana/assets/2946766/05d934ec-2d37-46d3-9a3b-b5c5933b6990">
